Reduce structured kernels' `set_output` boilerplate with new overloads.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77540
* #77539
* #76098
* #76097
* __->__ #76096

Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

ysiraichi committed:
Reduce structured kernels' `set_output` boilerplate with new overloads.

Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

Lezcano commented:
We should document somewhere (codebase or wherever there's the info about structured kernels) what each of these functions do, and what's their intended use.

ysiraichi commented:
@Lezcano Agreed. What do you think doing it right over the their signature?

Lezcano commented:
A short comment there surely wouldn't hurt. Now, it would be nice to have a note somewhere describing the whole thing. Do we have any note somewhere describing how structured kernels work @bdhirsh?  I remember there's the RFC, but that is (and will get more and more) out of date.

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -62,14 +77,58 @@ namespace impl {
+  virtual void set_output_raw_strided(
+      int64_t output_idx,
+      IntArrayRef sizes,
+      IntArrayRef strides,
```
I guess I'd suggest naming this something like `strides_hint` to indicate that you won't get this strides if an explicit out tensor is passed

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -62,14 +77,58 @@ namespace impl {
     set_output(output_idx, sizes, {}, options, {});
   }
+
+  // See: https://github.com/pytorch/pytorch/issues/69813
```
I feel the documentation here is not clear enough, people will not understand how to use this API correctly just by reading these comments. I think adding some sort of flow chart of how to decide pick which one will help.

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -13,6 +13,21 @@ namespace at {
+  }
+
+  return strides;
+}
```
I swear we should have another implementation of this lying around somewhere...

Lezcano replied:
In fact, not only we do, but this one happens to be incorrect! Strides should not be zero even when the sizes are zero.

I implemented this one, which also gives fortran-contiguous strides as a helper for linalg. It is based on how TI generates the strides.

https://github.com/pytorch/pytorch/blob/b17b2b1cc7b017c3daaeff8cc7ec0f514d42ec37/aten/src/ATen/native/LinearAlgebraUtils.h#L36-L66

Lezcano replied:
fwiw, I think we should put this implementation (or an equivalent one that works just for contiguous tensors) into a more visible header.

ysiraichi replied:
> I swear we should have another implementation of this lying around somewhere...

As far as I know, there are these 2:

- [`jit_type.h`](https://github.com/pytorch/pytorch/blob/0708630d9f30210b0a51d13558ad173da630ef95/aten/src/ATen/core/jit_type.h#L789-L798)
- [`LinearAlgebraUtils.h`](https://github.com/pytorch/pytorch/blob/0708630d9f30210b0a51d13558ad173da630ef95/aten/src/ATen/native/LinearAlgebraUtils.h#L36-L66)

> fwiw, I think we should put this implementation (or an equivalent one that works just for contiguous tensors) into a more visible header.

Agreed. That said, I'm not sure where to put it myself. This was the main reason I created one here.
Anyway, I will try to find a better place to leave it.

Lezcano replied:
This is general enough that it could be somewhere in c10, I'd say.

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -489,21 +529,26 @@ def gen_class_ctor(self, k: SchemaKind, class_name: str, returns: int) -> str:
         elif k is SchemaKind.out:
             output_type = "std::reference_wrapper<Tensor>"
+            output_value = "proxy_outputs_[output_idx].has_value() ? **proxy_outputs_[output_idx] : outputs_[output_idx].get()"
+            proxy_field = f"std::array<c10::optional<c10::ExclusivelyOwned<Tensor>>, {len(f.func.returns)}> proxy_outputs_;"
```
ExclusivelyOwned nice

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -489,21 +529,26 @@ def gen_class_ctor(self, k: SchemaKind, class_name: str, returns: int) -> str:
             out_refs = ', '.join(f"std::ref(out{i})" for i in range(returns))
-            return f"{class_name}({out_args}) : outputs_{{ {out_refs} }} {{}}"
+            nullopts = ", ".join("c10::nullopt" for _ in range(returns))
+            return f"{class_name}({out_args}) : outputs_{{ {out_refs} }}, proxy_outputs_{{ {nullopts} }} {{}}"
```
surprised this doesn't just default initialize appropriately

ysiraichi replied:
Hmmm. You are right! It does initialize appropriately. Will fix it.

ezyang commented:
This looks good. I'd like to see the generated code before merging.

Lezcano commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -461,6 +489,16 @@ def gen_class_set_output_body(self, k: SchemaKind) -> str:
+        if maybe_create_proxy:
+            create_proxy = """
+auto maybe_proxy = maybe_create_proxy(out, sizes, strides, options);
+if (maybe_proxy.has_value()) {
```
Nit. This might better be around a `C10_UNLIKELY`? The user is barely ever going to provide an out with the wrong strides.

ezyang commented:
bumping

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi commented:
Here's what the generated code for `add.Tensor` looks like (in `RegisterCPU.cpp`):

```c++
c10::optional<Tensor> maybe_create_proxy(const Tensor &out, IntArrayRef sizes, IntArrayRef strides, const TensorOptions &options) {
  if (out.strides() != strides) {
    return at::detail::empty_strided_cpu(sizes, strides, options);
  }
  return c10::nullopt;
}
```

```c++
struct structured_ufunc_add_CPU_out final : public at::native::structured_ufunc_add_CPU {
    structured_ufunc_add_CPU_out(Tensor& out0) : outputs_{ std::ref(out0) } {}


    void set_output_strided(
        int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
        TensorOptions options, DimnameList names
    ) override {

        const auto& out = outputs_[output_idx].get();
        resize_out(out, sizes, strides, options);

        auto maybe_proxy = maybe_create_proxy(out, sizes, strides, options);
        if (C10_UNLIKELY(maybe_proxy.has_value())) {
            proxy_outputs_[output_idx] = c10::ExclusivelyOwned<Tensor>(std::move(maybe_proxy).value());
        }

        if (!names.empty()) {
          namedinference::propagate_names(outputs_[output_idx], names);
        }
        // super must happen after, so that downstream can use maybe_get_output
        // to retrieve the output
        at::native::structured_ufunc_add_CPU::set_output_raw_strided(output_idx, sizes, strides, options, names);
    }


    void set_output_raw_strided(
        int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
        TensorOptions options, DimnameList names
    ) override {

        const auto& out = outputs_[output_idx].get();
        resize_out(out, sizes, strides, options);

        if (!names.empty()) {
          namedinference::propagate_names(outputs_[output_idx], names);
        }
        // super must happen after, so that downstream can use maybe_get_output
        // to retrieve the output
        at::native::structured_ufunc_add_CPU::set_output_raw_strided(output_idx, sizes, strides, options, names);
    }


    const Tensor& maybe_get_output(int64_t output_idx) override {
      return proxy_outputs_[output_idx].has_value() ? **proxy_outputs_[output_idx] : outputs_[output_idx].get();    }
    std::array<std::reference_wrapper<Tensor>, 1> outputs_;
    std::array<c10::optional<c10::ExclusivelyOwned<Tensor>>, 1> proxy_outputs_;
};
```
```c++
at::Tensor & wrapper_add_out_out(const at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha, at::Tensor & out) {
structured_ufunc_add_CPU_out op(out);
op.meta(self, other, alpha);
op.impl(self, other, alpha, op.maybe_get_output(0));
if (op.proxy_outputs_[0].has_value()) op.outputs_[0].get().copy_(**op.proxy_outputs_[0]);
return out;
}
```

ezyang reviewed (approved):
It's a little unfortunate that we're generating all of the code to handle proxies even for kernels where we *know* that the meta function is only ever going to call, e.g., `set_output_raw_strided` (and it's not just dead code, but you have to do the test to see if the proxy is set on the output, so that's an extra condition on the fastpath), but to apply this optimization we are going to have to add some extra metadata to `native_functions.yaml` so I guess it is fine to leave it out for now.

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Restored ATen/core/DimVector.h on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Update on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ezyang commented:
happy to merge this whenever it's green. LMK if there are substantive changes

ysiraichi committed:
Fix curly bracket. on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi commented:
Hmm. This CI failure doesn't look like it's related to this PR.

```
FAILED: /var/lib/jenkins/workspace/xla/build/temp.linux-x86_64-3.7/torch_xla/csrc/aten_xla_bridge.o 
clang++-8 -MMD -MF /var/lib/jenkins/workspace/xla/build/temp.linux-x86_64-3.7/torch_xla/csrc/aten_xla_bridge.o.d -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -D_GLIBCXX_USE_CXX11_ABI=1 -fPIC -I/var/lib/jenkins/workspace/xla -I/var/lib/jenkins/workspace/xla/third_party/tensorflow/bazel-tensorflow -I/var/lib/jenkins/workspace/xla/third_party/tensorflow/bazel-bin -I/var/lib/jenkins/workspace/xla/third_party/tensorflow/bazel-tensorflow/external/protobuf_archive/src -I/var/lib/jenkins/workspace/xla/third_party/tensorflow/bazel-tensorflow/external/com_google_protobuf/src -I/var/lib/jenkins/workspace/xla/third_party/tensorflow/bazel-tensorflow/external/eigen_archive -I/var/lib/jenkins/workspace/xla/third_party/tensorflow/bazel-tensorflow/external/com_google_absl -I/var/lib/jenkins/workspace -I/var/lib/jenkins/workspace/torch/csrc -I/var/lib/jenkins/workspace/torch/lib/tmp_install/include -I/opt/conda/lib/python3.7/site-packages/torch/include -I/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.7/site-packages/torch/include/TH -I/opt/conda/lib/python3.7/site-packages/torch/include/THC -I/opt/conda/include/python3.7m -c -c /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_bridge.cpp -o /var/lib/jenkins/workspace/xla/build/temp.linux-x86_64-3.7/torch_xla/csrc/aten_xla_bridge.o -std=c++14 -Wno-sign-compare -Wno-deprecated-declarations -Wno-return-type -Wno-macro-redefined -Wno-return-std-move -DNDEBUG -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_clang"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1002"' -DTORCH_EXTENSION_NAME=_XLAC -D_GLIBCXX_USE_CXX11_ABI=1
In file included from /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_bridge.cpp:11:
/var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:33:40: error: only virtual member functions can be marked 'override'
  at::IntArrayRef sizes_custom() const override;
                                       ^~~~~~~~
/var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:34:42: error: only virtual member functions can be marked 'override'
  at::IntArrayRef strides_custom() const override;
                                         ^~~~~~~~
/var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:36:30: error: only virtual member functions can be marked 'override'
  int64_t dim_custom() const override;
                             ^~~~~~~~
/var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:38:32: error: only virtual member functions can be marked 'override'
  int64_t numel_custom() const override;
                               ^~~~~~~~
```

ysiraichi committed:
Rebased. on "Reduce structured kernels' `set_output` boilerplate with new overloads."


Partially fix #69813
This PR does mainly 3 things:

1. Introduces new methods for the `MetaBase` API:
    - `set_output_strided`: creates proxy tensors with exact strides, if strides don't match
    - `set_output_contiguous`: alias for `set_output_strided` with contiguous strides
    - `set_output_raw_strided`: does not create proxy tensors

2. Modifies codegen for handling proxy tensors:
    - Creates a new field for out-of-place kernels: `proxy_output_`
    - Implements `set_output_strided` by creating a proxy tensor if necessary
    - Passes the proxy tensor to them `IMPL` function
    - Copy the result back to the real output, in the end, whenever a proxy was created

3. Replace `set_output` by `set_output_raw_strided` for `TensorIterator*`
    - Needed, since it overrides `set_output`

[ghstack-poisoned]

ysiraichi commented:
@ezyang 
The CI is already green, but it hasn't been merged. Could I send `merge this`?

ezyang commented:
@pytorchbot merge this
